### PR TITLE
Document option to allow almost all printable ASCII characters in environment variables

### PR DESCRIPTION
**Pull request by HirazawaUi**

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP]: Allow almost all printable ASCII characters in environment variables

**Comment by sftim** [comment]
/retitle [WIP]: Document option to allow almost all printable ASCII characters in environment variables

**Comment by fsmunoz** [comment]
@HirazawaUi I think that the target branch should be `1.30` instead of `main`, since this is related to a feature with a `1.30` milestone - would it be possible to change it?

**Comment by HirazawaUi** [comment]
> @HirazawaUi I think that the target branch should be `1.30` instead of `main`, since this is related to a feature with a `1.30` milestone - would it be possible to change it?

Ah, sorry, that was my mistake.

**Comment by HirazawaUi** [comment]
@sftim May you take a look at this document first? Because you have the full context and understand it well enough.

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (stmcginnis) [inline content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:36]: I think a lot of readers won't know what `C_IDENTIFIER` means and could be confused by this statement.

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (stmcginnis) [inline content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:37]: I would recommend removing line 36 and rewording this note.

```suggestion
If the `RelaxedEnvironmentVariableValidation` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled, all [printable ASCII characters](https://www.ascii-code.com/characters/printable-characters) except "=" may be used for environment variable names.
```

Comment (stmcginnis) [reply content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:37]: Though is that accurate? There seems to be other printable characters that would be problematic as environment variables.

Comment (HirazawaUi) [reply content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:37]: Your description is very correct, we have done enough tests on it.

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (HirazawaUi) [reply content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:36]: I followed our comments in the api documentation.
ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/

Maybe they'll be confused, until they use a search engine....

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (sftim) [inline content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:1]: I recommend against using the `{{< note >}}` shortcode inside the description for a feature gate. Instead, signpost readers to relevant documentation.

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (sftim) [reply content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:36]: If our docs site doesn't explain what a `C_IDENTIFIER` means (it doesn't really), then we shouldn't refer to it.

Instead, we could add a small page about environment variables for containers (inside https://kubernetes.io/docs/concepts/configuration/) - just a stub for now, and then add a heading about allowed characters inside variable names.

**Review by sftim — CHANGES_REQUESTED** [review]
Thanks.

Please also amend https://kubernetes.io/docs/concepts/configuration/configmap/, https://kubernetes.io/docs/concepts/configuration/secret/ and https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/

and optionally https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (HirazawaUi) [reply content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:36]: all right.

File: content/en/docs/concepts/configuration/environment.md
Comment (HirazawaUi) [inline content/en/docs/concepts/configuration/environment.md:15]: I'm not sure if this is enough.

**Comment by HirazawaUi** [comment]
> Hello @HirazawaUi 👋 please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday March 12th 2024 18:00 PST. Thank you!

The current PR is already in Ready for review, I did not create the draft PR initially.....

**Comment by katcosgrove** [comment]
Hey @HirazawaUi! Just a reminder that this needs to be reviewed and ready to merge by [Docs Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#docs-freeze) on March 26, including a technical review. An [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) will be required if you cannot meet that deadline. Thank you!

cc: @thockin @mrunalp

**Comment by HirazawaUi** [comment]
> > Please also amend https://kubernetes.io/docs/concepts/configuration/configmap/, https://kubernetes.io/docs/concepts/configuration/secret/
> 
> Maybe we shouldn't modify the documentation of configmap and secret. Our changes only apply to using `envFrom` to reference the keys of configmap and secret as environment variable names. Before, they did not mention this part.
> 
> For me, describing this feature in `environment-variable-expose-pod-information` and `define-environment-variable-container` together is a duplication, in `define-environment-variable-container` we describe how to define an environment variable, and in `environment-variable-expose-pod-information` we describe how to expose pod information through environment variables, and based on semantics I prefer the current way(describe in define-environment-variable-container)
> 
> WDYT?

@sftim  Do you have any follow-up thoughts?

File: content/en/docs/concepts/configuration/environment.md
Comment (sftim) [inline content/en/docs/concepts/configuration/environment.md:2]: People will think that environment containers are a fourth kind alongside:
- sidecar containers
- init containers
- application containers

I would drop this page.

File: content/en/docs/tasks/inject-data-application/define-environment-variable-container.md
Comment (sftim) [inline content/en/docs/tasks/inject-data-application/define-environment-variable-container.md:41]: I (strongly) recommend putting this advice under its own heading, within the “Using environment variables inside of your config” heading. We should explain the default behavior (restricting what characters are allowed) and the experimental behavior that allows a wider range of characters.

**Comment by sftim** [comment]
> Maybe we shouldn't modify the documentation of configmap and secret. Our changes only apply to using `envFrom` to reference the keys of configmap and secret as environment variable names. Before, they did not mention this part.
> 
> For me, describing this feature in `environment-variable-expose-pod-information` and `define-environment-variable-container` together is a duplication, in `define-environment-variable-container` we describe how to define an environment variable, and in `environment-variable-expose-pod-information` we describe how to expose pod information through environment variables, and based on semantics I prefer the current way(describe in define-environment-variable-container)

In technical writing, we like to overcommunicate some details. We can't make assumptions about which page readers are looking at to learn about a thing, or which pages they have already read.

I recommend adding a heading and brief text into https://kubernetes.io/docs/concepts/configuration/configmap/#using-configmaps that mentions environment variables, and that links to one of the pages where this new feature is mentioned.
You may well only write 100 extra words.

I recommend adding a mention to https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables that the name of the environment variable is subject to restrictions. The way I'd do it, “restrictions” would be a hyperlink to some other page, eg to an updated version of `content/en/docs/tasks/inject-data-application/define-environment-variable-container.md`

**Review by sftim — COMMENTED** [review]
Thanks for this PR. We can't merge it as-is; I've added some feedback to help you revise it.

**Comment by sftim** [comment]
> @sftim May you take a look at this document first? Because you have the full context and understand it well enough.

I'm afraid we can't offer you that luxury; our SIG leads' time is popular. Please work with reviewers for English, me included, but not me exclusively. I wouldn't have the capacity to pick up all the reviews I've been invited to help out with ahead of the docs deadline.

**Comment by HirazawaUi** [comment]
> > @sftim May you take a look at this document first? Because you have the full context and understand it well enough.
> 
> I'm afraid we can't offer you that luxury; our SIG leads' time is popular. Please work with reviewers for English, me included, but not me exclusively. I wouldn't have the capacity to pick up all the reviews I've been invited to help out with ahead of the docs deadline.

Okay, I understand. Thank you for your patient comment. I will communicate directly with the English reviewer later.

**Comment by drewhagen** [comment]
Hello @liggitt @thockin a reminder that this needs a technical review before being ready to merge by [Docs Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#docs-freeze) on March 26. An [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) will be required if you cannot meet that deadline.

@kubernetes/sig-docs-en-reviews I see there is request for additional review (see comments above). Might someone other than Tim be available to assist review on this one?

Thanks everyone!

**Comment by HirazawaUi** [comment]
Before reviewing by sig-docs-en-reviews, I would like to hear @thockin opinion on [comment](https://github.com/kubernetes/kubernetes/pull/123385#discussion_r1527139998). We are allowed to relax the restrictions on using EnvFrom to reference configmap/secret keys when feature gate is not enable.

**Comment by Princesso** [comment]
Hello @HirazawaUi  :wave:! I'm reaching out from the Docs team. Just checking in as we approach Docs Freeze on Tuesday March 26th 18:00 PDT. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as lgtm and approve labels applied, without any unaddressed comments or concerns from SIG Docs. The status of this enhancement is marked as at risk for docs freeze. Thank you!

**Comment by HirazawaUi** [comment]
> Hello @HirazawaUi 👋! I'm reaching out from the Docs team. Just checking in as we approach Docs Freeze on Tuesday March 26th 18:00 PDT. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as lgtm and approve labels applied, without any unaddressed comments or concerns from SIG Docs. The status of this enhancement is marked as at risk for docs freeze. Thank you!

OK, I'm also waiting for the technical review. If @thockin doesn't have any free time, I will try to ask other approvers of sig-node for help.

**Comment by natalisucks** [comment]
@kubernetes/sig-node-leads Howdy leads, we're looking for a technical review for these docs to make the Docs Freeze deadline of March 26 – thanks!

File: content/en/docs/concepts/configuration/secret.md
Comment (thockin) [inline content/en/docs/concepts/configuration/secret.md:568]: Is this URL line-wrapped?

File: content/en/docs/concepts/configuration/configmap.md
Comment (thockin) [inline content/en/docs/concepts/configuration/configmap.md:240]:
```suggestion
It's important to note that the range of characters allowed for environment
variable names in pods is
[restricted](/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config),
If any keys do not meet the rules, those keys are not made available to your container, though
the Pod is allowed to start.
```

**Review by thockin — COMMENTED** [review]
Technical review LGTM

**Review by mrunalp — APPROVED** [review]
/approve
Sig-node

**Comment by drewhagen** [comment]
Just a heads up that Doc Freeze is approaching fast tomorrow March 26th at 18:00 PDT and starting this cycle, we need to file [an Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) after this time.
This one looks pretty close and we'd like to help you get it over the finish line!

@kubernetes/sig-docs-en-owners Does this one look good to you? @HirazawaU I notice we have approvals from SIG Node leads and also there are a couple outstanding comments from Tim H. Do we want to tackle those before merge?

**Comment by HirazawaUi** [comment]
> Just a heads up that Doc Freeze is approaching fast tomorrow March 26th at 18:00 PDT and starting this cycle, we need to file [an Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) after this time. This one looks pretty close and we'd like to help you get it over the finish line!
> 
> @kubernetes/sig-docs-en-owners Does this one look good to you? We could use an `approve` @HirazawaUi I notice we have approvals from SIG Node leads and also there are a couple outstanding comments from Tim H. Do we want to tackle those before merge?

All comments have been fixed.

@kubernetes/sig-docs-en-owners for lgtm and approve

**Review by divya-mohan0209 — COMMENTED** [review]
@drewhagen/@HirazawaUi: I'd recommend we check if the links for this one are wrapped properly. Other than that, I'll bias for action here.
/approve

File: content/en/docs/concepts/configuration/secret.md
Comment (HirazawaUi) [reply content/en/docs/concepts/configuration/secret.md:568]: Sorry for my bad English, I misunderstood `line-wrapped`, it does contain line wrap, it has been fixed in #45670.

**Pull request merged**
